processing/src/main/java/org/apache/druid/math/expr/BinaryLogicalOperatorExpr.java

```diff
@@ -331,7 +331,9 @@ public ExprEval eval(ObjectBinding bindings)
       return ExprEval.ofLongBoolean(false);
     }
     ExprEval rightVal;
-    if (NullHandling.sqlCompatible() || Types.is(leftVal.type(), ExprType.STRING)) {
+    // null values can (but not always) appear as string typed
+    // so type isn't necessarily string unless value is non-null
+    if (NullHandling.sqlCompatible() || (Types.is(leftVal.type(), ExprType.STRING) && leftVal.value() != null)) {
```

Comment:
I feel like I'm not understanding this logic, but it seems like previously you were doing null checks if it was a String.  Now,  you will only do null checks if it is a String *and* the left-hand side is not null?  This seems to be trying to coerce for boolean operations, if the left side is numerical but also not null, it seems like we should have semantics that are like `if == 0 then false, else true`.  But, I think that this will just ignore it if it's not String?

Reply:
I initially made this change before i added `valueOrDefault`, because in a situation where expressions are evaluated without the type information to tell us that a null is a long (such as ingestion time transforms), an expression like `null && 1`  with the left side is a null value it will be evaluated as a `STRING` typed expression, and follow the string/sql compatible rules, in default value mode meaning `value()` spits out `null`. if the expression was `1 && null` however, it would be evaluated with long rules, and spit out `0` in default value mode.

None of this really matters now though, since callers should be using `valueOrDefault()` after the changes in this patch, instead of `value()`, which produces equivalent results, so I will switch it back and switch the tests around null constants to use that.